text/0088-fix-memory-limitiations-in-session-replays-access-pattern.md

```diff
@@ -0,0 +1,232 @@
+
+**Drawbacks**
+
+There are aggregated values which users want to sort and filter on. Columns such as count_errors and activity which are prominently featured on our index page. Sorting and filtering by these values disables the preflight query and only runs the aggregated query.
+
+The non-aggregated query does not allow **exclusive** filter conditions against **non-static** columns. For example, we can not say "find a replay where this url does not exist". The query will find _rows_ where that condition is true but it will not find _replays_ where that is condition true.
+
+The non-aggregated query does not allow multiple, **inclusive** filter conditions against **non-static** columns. For example, we can not say "find a replay where this url exists and this other url exists". It will find **ROWS** which have both urls but not **REPLAYS** which have both urls. Transforming the `AND` operator to an `OR` operator does not satisfy the condition because it will match replays which contain one of the urls - not both.
+
+# Options Considered
+
+Any option may be accepted in whole or in part. Multiple options can be accepted to achieve the desired outcome. The options are ordered from perceived ease to perceived difficulty.
+
+### 1. Change the Product's Query Access Pattern
+
+**Proposal**
+
+Our current subquery solution works very well. However, there are escape hatches which require us to issue an aggregation query over the whole dataset. Those escape hatches should be closed.
+
+- Remove ability to filter and sort by aggregated values (e.g. count_errors, activity_score).
+- Remove ability to filter by negation values which change (e.g. urls).
+- Remove ability to use the AND operator when a non-static value is present.
+  - Filtering by browser_name AND os_name would be permitted.
+  - Filtering by browser_name AND click_action would be disallowed.
+    - Click rows do not store browser information because they are internally generated.
+    - Should this condition change these filters could be supported.
+  - Filtering by error_id AND url would be disallowed.
+    - Filtering by error_id OR url would be allowed.
+
+**Drawbacks**
+
+- Significant reduction in index page quality.
+- Removes the most important features of our index page. Count-errors sort, activity sort, duration sort and filter.
+
+**Questions**
+
+### 2. Reduce the Scan Range on the Front End
+
+**Proposal**
+
+The number of rows aggregated can be reduced by restricting the maximum time range we query over. We should validate the timestamp range such that it does not exceed a 24-hour period. This would satify every organization which ingests fewer than 1 billion replay-segments every 90 days.
+
+**Drawbacks**
+
+- To search for a unique value a user would need to issue a query for each day of the retention period or until the value was found.
+- Small customers would not see any replays on their index page due to the limited window size.
```

Comment:
I think you could probably issue multiple queries in this case.

1) Fetch a small initial window (past 24 hours).
2) If you get more than N results, stop.
3) If you get less than N results, query again for another time window (last week - last 24 hours).
4) Repeat 3 until you get N results.